pkg/scheduler/api/compatibility/compatibility_test.go

```diff
@@ -1218,34 +1224,48 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
 		if !reflect.DeepEqual(policy, tc.ExpectedPolicy) {
 			t.Errorf("%s: Expected:\n\t%#v\nGot:\n\t%#v", v, tc.ExpectedPolicy, policy)
 		}
-
+		policyConfigMap := v1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{Namespace: metav1.NamespaceSystem, Name: "scheduler-custom-policy-config"},
+			Data:       map[string]string{schedulerconfig.SchedulerPolicyConfigMapKey: tc.JSON},
+		}
+		policyConfigMap.APIVersion = "v1"
+		body, err := json.Marshal(policyConfigMap)
+		if err != nil {
+			t.Errorf("Error creating configMap: %+v", err)
+		}
 		handler := utiltesting.FakeHandler{
 			StatusCode:   500,
-			ResponseBody: "",
+			ResponseBody: string(body),
 			T:            t,
 		}
 		server := httptest.NewServer(&handler)
 		defer server.Close()
 		client := clientset.NewForConfigOrDie(&restclient.Config{Host: server.URL, ContentConfig: restclient.ContentConfig{GroupVersion: &schema.GroupVersion{Group: "", Version: "v1"}}})
+		algorithmSrc := schedulerconfig.SchedulerAlgorithmSource{
+			Policy: &schedulerconfig.SchedulerPolicySource{},
```

Comment:
shouldn't this specify config map as the algorithm source? The tests should fail without that I believe.

Reply:
That's because this test is using FakeHandler (line 1236) which was updated to return the ConfigMap. Depending on the server to get the config map won't work in this case (because of the FakeHandler).

Reply:
I don't get it, scheduler.New below will try to setup the scheduler using ```algorithmSrc```, which seems to me half initialized: it specifies Policy (vs algorithm provider), but does not specify the source of Policy (ConfigMap vs File). I am not sure how this is related to how ConfigMap is served.

Reply:
Aha, I got the point now. We should have an empty ConfigMap inside the Policy as well .. Am I correct?

Reply:
you need to populate SchedulerPolicyConfigMapSource with the name and namespace of the configmap

Reply:
I am surprised this is passing the tests as it is...

Reply:
Me, not so much. This test is just testing the parsing of the JSON into a policy, instead of the effects in the scheduler. This is wrong, of course.

Reply:
@alculquicondor but the scheduler did not return any error. empty policy is ok in scheduler creation, nothing checks for it.

Reply:
Right, but that means that we are not using the policy that we want to use :)

Reply:
sure, just wanted to point out a reason why the test passed.

Reply:
is there a way to check that the policy the scheduler is using is the expected one?

Reply:
> Me, not so much. This test is just testing the parsing of the JSON into a policy, instead of the effects in the scheduler. This is wrong, of course.

@alculquicondor one would have to run the scheduler and schedule at least one pod to validate the policy can produce a scheduler capable of scheduling. Though it will still not answer the question if the policy is really usable. There are (or should be) other tests validating scheduler's capabilities given a list of policies.

> is there a way to check that the policy the scheduler is using is the expected one?

Are you asking for (new line 1224):

```go
if !reflect.DeepEqual(policy, tc.ExpectedPolicy) {
	t.Errorf("%s: Expected:\n\t%#v\nGot:\n\t%#v", v, tc.ExpectedPolicy, policy)
}
```
?

Reply:
> I am surprised this is passing the tests as it is...

@ahg-g unfortunate constellation. `pkg/api/scheduler/api.validation.ValidatePolicy` considers empty `schedulerapi.Policy{}` as valid (everything falls down to defaults). Also:

```go
case source.Policy != nil:
	// Create the config from a user specified policy source.
	policy := &schedulerapi.Policy{}
	switch {
	...
	case source.Policy.ConfigMap != nil:
		if err := initPolicyFromConfigMap(client, source.Policy.ConfigMap, policy); err != nil {
			return nil, err
		}
	}
```

the inner case block does not check if `source.Policy.ConfigMap.Namespace`, resp. `source.Policy.ConfigMap.Name` is set since in relies on the clientset to tell if it exists or not.

Instead, we can use fake clientset when initializing the client:
```go
client := fake.NewSimpleClientset(&policyConfigMap)
```

which will provided the expected error message as the fake client will not find CM with an empty name.

Reply:
Surprise, surprise. In case there is only a single object added into the fake client, setting both namespace and name  to an empty string "will" return the object. One might temporarily create another CM:
```
client := fake.NewSimpleClientset(
	&policyConfigMap,,
	&v1.ConfigMap{ObjectMeta: metav1.ObjectMeta{Name: "test"}},
)
```

Upstream issue reported here: https://github.com/kubernetes/client-go/issues/685

Reply:
I think we could validate that the policy was applied correctly by calling `scheduler.Algorithm.Predicates()` and `scheduler.Algorithm.Prioritizers()`.

Reply:
@ingvagabund I believe, this wouldn't be a problem in our case, as the test is creating a single policyConfigMap. Am I correct?

Reply:
>I think we could validate that the policy was applied correctly by calling scheduler.Algorithm.Predicates() and scheduler.Algorithm.Prioritizers()

How would we compare those to priorities (they are of different types)? should we only compare names?

Reply:
Today, this test compares the parsing of the json into an expected policy. What we should do is remove that expected policy and have expected predicates and expected prioritizers.
FYI, in go, prefer the word `want` instead of `expected` for the variables' names.

Reply:
@alculquicondor what about extenders? ExtenderConfigs are not exposed by `ScheduleAlgorithm`.

Reply:
They seem to be part of Algorithm as well (just not exposed)
https://github.com/kubernetes/kubernetes/blob/a56e86a405a414f83e5bb1d4e84e37d80af543b7/pkg/scheduler/factory/factory.go#L416

We can simply add the interface.